### PR TITLE
Move `Statement` and `Expr` so they hold identifiable structs

### PR DESCRIPTION
**Pull request by tvallotton**
This is of course a major breaking change. But I think it's worth considering. I also fixed the `DROP CONSTRAINT` thing I mentioned in #342.

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @tvallotton  👍 

In general,  I don't feel qualified to approve / accept this level of change without some additional buy in from @nickolay  / @andygrove  / @Dandandan . https://github.com/sqlparser-rs/sqlparser-rs/issues/311#issuecomment-902861958 seems like a good usecase to me but it is a pretty major change for all users of sqlparser-rs.

I view myself as a caretaker now and feel good about merging and releasing "round out the SQL support" type PRs. Splitting out the changes to `ALTER TABLE` as a separate PR, for example, would be fine,

**Comment by alamb** [comment]
See also https://github.com/sqlparser-rs/sqlparser-rs/issues/311

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
```
[the remark this answers is not in the document]
Comment (tvallotton) [reply .vscode/settings.json:1]: Oh yeah I think I was a little careless as I initially didn't think I was going to submit a pull request. I also noticed I made the changes on the master branch, would you prefer I did another submission on a separate branch?

[the remark this answers is not in the document]
Comment (alamb) [reply .vscode/settings.json:1]: > I also noticed I made the changes on the master branch, 

I don't think it matters what branch you work on in your fork (`tvallotton:main` is fine)

**Comment by alamb** [comment]
Closing old seemingly stalled PR. Please re-open if this was a mistake

**Comment by alamb** [comment]
Sadly I don't have the time to help drive projects such as this. Unless one of the other maintainers wants to do so, PRs like this are not likely to be merged. I am sorry I don't have the time and I am sorry if your time was wasted. 

As the current state of affairs might not have been clear in the project's README. I  have added  https://github.com/sqlparser-rs/sqlparser-rs/pull/416 to try and help

**Pull request closed without merging**